### PR TITLE
gh-94924: support `inspect.iscoroutinefunction` in `create_autospec(async_def)`

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-94924

### CODE REVIEW

**Comment by graingert** [comment]
see also https://github.com/python/cpython/issues/84753

File: Lib/unittest/mock.py [hunk 1]
```diff
@@ -171,7 +171,7 @@ def _instance_callable(obj):
     return False
 
 
-def _set_signature(mock, original, instance=False):
+def _set_signature(mock, original, instance=False, is_async_mock=False):
```
Comment (graingert) [inline Lib/unittest/mock.py]:
```suggestion
def _set_signature(mock, original, instance=False):
```

File: Lib/unittest/mock.py [hunk 2]
```diff
@@ -197,6 +197,33 @@ def checksig(*args, **kwargs):
     _setup_func(funcopy, mock, sig)
     return funcopy
 
+def _set_async_signature(mock, original, instance=False, is_async_mock=False):
+    # creates an async function with signature (*args, **kwargs) that delegates to a
```
Comment (akulakov) [inline Lib/unittest/mock.py:201]: This should probably be a docstring?

**Review by cjw296 — APPROVED** [review]

File: Lib/unittest/mock.py [hunk 3]
```diff
@@ -204,6 +204,33 @@ def checksig(*args, **kwargs):
...
+    # mock. It still does signature checking by calling a lambda with the same
+    # signature as the original.
+
+    skipfirst = isinstance(original, type)
+    result = _get_signature_object(original, instance, skipfirst)
+    if result is None:
+        return mock
```
Comment (cjw296) [inline Lib/unittest/mock.py:215]: @graingert - this path has no test coverage, please can you put in a PR to add some?

File: Lib/unittest/mock.py [hunk 4]
```diff
@@ -204,6 +204,33 @@ def checksig(*args, **kwargs):
...
+    func, sig = result
+    def checksig(*args, **kwargs):
+        sig.bind(*args, **kwargs)
+    _copy_func_details(func, checksig)
+
+    name = original.__name__
+    if not name.isidentifier():
+        name = 'funcopy'
```
Comment (cjw296) [inline Lib/unittest/mock.py:223]: @graingert - same with this branch, see: 
https://app.circleci.com/pipelines/github/testing-cabal/mock/1428/workflows/7b4dab2f-46de-4dd7-bb8b-5d2000975547/jobs/20351

(and https://github.com/testing-cabal/mock/pull/510)

File: Lib/unittest/mock.py (hunk 3 above)
Comment (cjw296) [reply Lib/unittest/mock.py:215]: https://github.com/python/cpython/pull/106617

File: Lib/unittest/mock.py (hunk 4 above)
Comment (cjw296) [reply Lib/unittest/mock.py:223]: https://github.com/python/cpython/pull/106617

**Comment by cjw296** [comment]
I think the issue can be closed, if folks need this in 3.12 or 3.11 they can comment here and I'll crank the handle to get it into the backport release on pypi.

**Pull request merged**